akka-stream/src/main/scala/akka/stream/impl/TraversalBuilder.scala

```diff
@@ -0,0 +1,985 @@
+
+  override def unwiredOuts: Int = if (outPort.isDefined) 1 else 0
+
+  def append[A, B, C](toAppend: TraversalBuilder, shape: Shape, matCompose: (A, B) ⇒ C): LinearTraversalBuilder =
+    append(LinearTraversalBuilder.fromBuilder(toAppend, shape, Keep.right), matCompose)
+
+  // We don't really need the Shape for the linear append, but it is nicer to keep the API uniform here
+  /**
+   * Append any builder that is linear shaped (have at most one input and at most one output port) to the
+   * end of this graph, connecting the output of the last module to the input of the appended module.
+   */
+  def append[A, B, C](toAppend: LinearTraversalBuilder, matCompose: (A, B) ⇒ C): LinearTraversalBuilder = {
+
+    if (toAppend.isEmpty) this
+    else if (this.isEmpty) {
+      toAppend.copy(
+        traversalSoFar = toAppend.traversalSoFar.concat(LinearTraversalBuilder.addMatCompose(traversal, matCompose))
+      )
+    } else {
+      if (outPort.nonEmpty) {
+        require(toAppend.inPort.isDefined, "Appended linear module must have an unwired input port " +
+          "because there is a dangling output.")
+
+        val traversalWithWiringCorrected = pendingBuilder match {
+          case Some(composite) ⇒
+            val out = outPort.get
+            applyAttributes(
+              composite
+                .assign(out, -composite.offsetOfModule(out) - toAppend.inSlots + toAppend.inOffset)
+                .traversal
+                .concat(traversalSoFar)
+            )
+          case None ⇒
+            // No need to rewire if input port is at the expected position
+            if (toAppend.inOffset == (toAppend.inSlots - 1))
+              traversal
+            else
+              applyAttributes(rewireLastOutTo(traversalSoFar, toAppend.inOffset - toAppend.inSlots))
+        }
+
+        val newTraversal =
+          if (toAppend.pendingBuilder.isEmpty)
+            toAppend.traversal
+              .concat(LinearTraversalBuilder.addMatCompose(traversalWithWiringCorrected, matCompose))
+          else {
+            toAppend.traversalSoFar
+              .concat(PopAttributes)
+              .concat(LinearTraversalBuilder.addMatCompose(traversalWithWiringCorrected, matCompose))
+          }
+
+        // Simply just take the new unwired ports, increase the number of inSlots and concatenate the traversals
+        LinearTraversalBuilder(
+          inPort = inPort,
+          outPort = toAppend.outPort,
+          inSlots = inSlots + toAppend.inSlots,
+          inOffset = inOffset + toAppend.inSlots,
+          // Build in reverse so it yields a more efficient layout for left-to-right building
+          traversalSoFar = newTraversal,
+          pendingBuilder = toAppend.pendingBuilder,
+          attributes = Attributes.none,
+          beforeBuilder = if (toAppend.pendingBuilder.isEmpty) EmptyTraversal else PushAttributes(toAppend.attributes)
+        )
+      } else throw new Exception("should this happen?")
```

Comment:
no, this should never happen :) Make this an explicit requirement and throw a proper exception when not met.